### PR TITLE
Random album access version 1

### PR DESCRIPTION
**Pull request by Octoton**
First part of the perpetual queue implementation following queue revamp. For this version only album (hardcoded) are shufflelabe and there is no settings to focus the random search on specific thing (to lighten the PR size)

When activated a next album will be randomly choosen and display at the end of the queue, you can ask for another one with some choice or stop the perpetual queue:
Screenshot_20211230-212009  
  
There is more than one way to start a new perpetual queue:
Screenshot_20211230-212004
Screenshot_20211230-211955
Screenshot_20211230-211931  
  
On other screen than the album one the classic shuffle song is shown (if we add other shuffling like by artist for example, we can change the same way than the album tab the artist tab):
Screenshot_20211230-211926

There is a settings to remember or not if the perpetual queue was activated or not:
Screenshot_20211230-212340

### CODE REVIEW

**Comment by Octoton** [comment]
@soncaokim there is the first of two PR to implement a random album shuffling with easy way to have other type in the futur

**Comment by soncaokim** [comment]
Hey, sorry I did not have time nor courage to answer you. 

I just have an extreme quick look, and havent noticed anything blocking. The concept of DynQueue appears in core classes, so this should be kept minimal.

Is this stable enough for me testing? Or I should stay within the safe distance for now?

**Comment by Octoton** [comment]
@soncaokim I did this mounth ago and I haven't have problem with it ever since (code it just after static queue, but got side track on other stuff. Well better late than never i guess)
There is a next PR with ensure no constant repeat and less random album each time, but this PR should work nonetheless.

**Comment by soncaokim** [comment]
- Can you please solve the (minor) merge conflict?
- The changes on core `*Fragment` and `*Activity` classes bother me a bit, since they need to be aware of the underlying queue implementation to do refresh/reload. I dont have any concrete counter proposition though.
- If one wants to propose a new dynqueue idea, the **only** change on core classes would be `MusicService.restoreQueuesAndPositionIfNecessary` correct ?

**Comment by Octoton** [comment]
- Done (had to move the staticQueue correction you did in https://github.com/AdrienPoupa/VinylMusicPlayer/pull/547, see new commit)
- Me too :), but I don't see any really clean way to do this
- More like MusicService.setQueueToDynamicQueue, restoreQueuesAndPositionIfNecessary is there to restore an already created queue. But yes, the idea is that when you change the QueueLoader associated with the dynamic queue, everything change (this will be more clear with the next PR).

File: app/src/main/java/com/poupa/vinylmusicplayer/misc/queue/DynamicElement/AlbumShuffling/DB.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    private static final String DATABASE_NAME = "shuffling_album.db";
+    private static final int VERSION = 1;
+
+    DB() {
+        super(App.getInstance().getApplicationContext(), DATABASE_NAME, null, VERSION);
+    }
+
+    @Override
+    public void onCreate(@NonNull final SQLiteDatabase db) {
+        /*db.execSQL("CREATE TABLE IF NOT EXISTS " + ListenHistoryColumns.NAME + " ("
```
Comment (soncaokim) [inline app/src/main/java/com/poupa/vinylmusicplayer/misc/queue/DynamicElement/AlbumShuffling/DB.java:28]: Please drop the commented code if not needed (here and below)

File: app/src/main/java/com/poupa/vinylmusicplayer/misc/queue/DynamicElement/AlbumShuffling/DB.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+import com.poupa.vinylmusicplayer.App;
+import com.poupa.vinylmusicplayer.misc.queue.DynamicElement.DynamicElement;
+
+
+// For album shuffling V2: sqlite usage should be squash if possible + future album history is saved
+/**
+ * Provide saving of album used in {@link DynamicElement} of AlbumShuffling implementation,
+ * needed to ensure reopening of the app after a deep sleep will not result in a different playing queue state  */
+public class DB extends SQLiteOpenHelper {
+    private static final String DATABASE_NAME = "shuffling_album.db";
```
Comment (soncaokim) [inline app/src/main/java/com/poupa/vinylmusicplayer/misc/queue/DynamicElement/AlbumShuffling/DB.java:19]: I would add some prefix here to the DB name, i.e "dynqueue_"

**Comment by soncaokim** [comment]
- I'm under the impression that when this is enabled, the very first song of each album is not queued/not played.

- i dont think that the "stop perpetual queue" should be colored red, ie not an destructive action as the case of song deletion.

Screenshot_20220106-224224-544

File: app/src/main/java/com/poupa/vinylmusicplayer/misc/queue/DynamicElement/AlbumShuffling/DB.java (hunk 1 above)
Comment (Octoton) [reply app/src/main/java/com/poupa/vinylmusicplayer/misc/queue/DynamicElement/AlbumShuffling/DB.java]: 👍

File: app/src/main/java/com/poupa/vinylmusicplayer/misc/queue/DynamicElement/AlbumShuffling/DB.java (hunk 2 above)
Comment (Octoton) [reply app/src/main/java/com/poupa/vinylmusicplayer/misc/queue/DynamicElement/AlbumShuffling/DB.java]: 👍

**Comment by Octoton** [comment]
> * I'm under the impression that when this is enabled, the very first song of each album is not queued/not played.
> 
> * i dont think that the "stop perpetual queue" should be colored red, ie not an destructive action as the case of song deletion.

This is strange because for me the first song is always played, could you put the step by step way that you used to see this behavior ?  
  
More than destructive action, the red button inside a menu is there to quickly warn user with quick thumb that the action will be an hassle to revert => when you stop the perpetual queue by mistake and then begin a new one, next song will be different and in future implementation other stuff could be lost as it stop all dynamic services (for example in next PR, album listening history will be flush, thus you can have the an album that you just listen too that is proposed to you). 
In contrast to this, when you remove a song (only action that is close but not red) from a queue by mistake you can just add it back.

**Review by soncaokim — APPROVED** [review]
This looks good to me

**Comment by soncaokim** [comment]
@Octoton Can you please fix the merge conflict? Afterwards I think we can merge this, @AdrienPoupa?

**Comment by Octoton** [comment]
Hi, sort if forgot about this to be honest 😅
Will try to do this in the week

**Comment by Octoton** [comment]
@soncaokim  there, it should be ok to merge now

**Comment by Octoton** [comment]
?

**Comment by Octoton** [comment]
?

**Comment by soncaokim** [comment]
Hi @Octoton ,

Given there is no response/activity from @AdrienPoupa for a while, and there is no release since ~1y, I'm considering forking this repo. While there is no guarantee that I can commit more time than right now on it, at least I can have the control on the release process and eventually see what is blocking.

The ideal setup would be to turn this repo to a team repo (and not a personal one), but I dont know if Adrien is okay or if he even follow this anymore.

S.

**Comment by Octoton** [comment]
Seems like a great way to keep this project alive

**Comment by soncaokim** [comment]
> Given there is no response/activity from @AdrienPoupa for a while, and there is no release since ~1y, I'm considering forking this repo. While there is no guarantee that I can commit more time than right now on it, at least I can have the control on the release process and eventually see what is blocking.

Hi @Octoton, please have a look at this [forked repo](https://github.com/vinyl2-team/vinyl2) and this [board](https://github.com/orgs/vinyl2-team/projects/1/views/2) where I've outlined steps to reach a v2 release.

Everything is still draft, and moving. And there are works to do before getting to a release-able state.

I've setup CI via Github Actions (which reuses the same CircleCI signing key, so that the development binaries can be upgraded seamlessly).

If you think this is a good target, is that possible to port this PR to the new repo? There should be some merging conflict to solve, but I'll try to assist there. I think this is a good time window to merge our PR in, before I start large-scale change on the code (renaming, linting, refactoring).

**Comment by Octoton** [comment]
> > Given there is no response/activity from @AdrienPoupa for a while, and there is no release since ~1y, I'm considering forking this repo. While there is no guarantee that I can commit more time than right now on it, at least I can have the control on the release process and eventually see what is blocking.
> 
> Hi @Octoton, please have a look at this [forked repo](https://github.com/vinyl2-team/vinyl2) and this [board](https://github.com/orgs/vinyl2-team/projects/1/views/2) where I've outlined steps to reach a v2 release.
> 
> Everything is still draft, and moving. And there are works to do before getting to a release-able state.
> 
> I've setup CI via Github Actions (which reuses the same CircleCI signing key, so that the development binaries can be upgraded seamlessly).
> 
> If you think this is a good target, is that possible to port this PR to the new repo? There should be some merging conflict to solve, but I'll try to assist there. I think this is a good time window to merge our PR in, before I start large-scale change on the code (renaming, linting, refactoring).

Will look into this in the week 👍

**Comment by AdrienPoupa** [comment]
Hi @soncaokim, sorry for the inactivity. I would like to help you setup actions so a new app can be published to the Play Store from this repo, I will investigate this, but if you have pointers on how to do it I would gladly help!

**Comment by soncaokim** [comment]
> Hi @soncaokim, sorry for the inactivity, and thanks for all the work you have done. I would like to help you setup actions so a new app can be published to the Play Store from this repo, I will investigate this, but if you have pointers on how to do it I would gladly help!

Hey there, long time no see! 😀

Well, as I stated above, I've started a fork and aim to build a team repo instead of personal repo, ie having the infra to let (trusted) people backing up each others. If we have a team of few trusted people, that would improve the current situation.

I would like also to have a "branding" continuity, so would like to ask you to reuse/base off the Vinyl app icon, and to name the new app Vinyl2 (or Vinyl Remastered/Reissued).
Would that be OK for you?

And if you think this can be a good common target, then please come join me.

Son

**Comment by soncaokim** [comment]
About the automation: this topic will be definitely an important focus for the new repo. I've already had a CI based on GithubActions running, with more coming.

**Comment by AdrienPoupa** [comment]
You are of course free to fork this as you see fit, but I would think it would be more beneficial for the project and its user base to keep using this repository to avoid segregating the user base / downloads.

I am open to any idea you may have to reduce the bus factor (even though I am not planning to be run over in the near future ;))

I am not opposed to transferring the repo to an organization, if you feel this is a step in the right direction, but I feel it's important to keep the user base / stars / downloads for the continuity of the project.

**Comment by soncaokim** [comment]
> I am not opposed to transferring the repo to an organization, if you feel this is a step in the right direction, but I feel it's important to keep the user base / stars / downloads for the continuity of the project.

That'll be a nice move if achievable. I'm afraid that this will imply sharing some secrets as well (release key, etc)

**Comment by soncaokim** [comment]
> I am open to any idea you may have to reduce the bus factor (even though I am not planning to be run over in the near future ;))

Wha, what happened? If you mean it literally, then "bon retablissement"

**Comment by AdrienPoupa** [comment]
> > I am open to any idea you may have to reduce the bus factor (even though I am not planning to be run over in the near future ;))
> 
> Wha, what happened? If you mean it literally, then "bon retablissement"

Nothing haha, I just meant it is better to have an organization so there is no single point of failure.

You will notice I created the VinylMusicPlayer and transferred the repo there. It has the benefits of keeping all the issues and stars, the former URL still works. I made you an admin of the organization so hopefully I am not a bottleneck anymore.

The last thing I see we still need to do it create an action to automate Play Store deployment, and possibly update FDroid metadata with the new repo URL.

Let me know if you need anything else!

**Comment by soncaokim** [comment]
> The last thing I see we still need to do it create an action to automate Play Store deployment, and possibly update FDroid metadata with the new repo URL.
> 
> Let me know if you need anything else!

Thanks, I'll have a look. If things are okay over there, I'll push a bunch of PR from my fork.
Let's continue over there, this thread is polluted with this discussion :-)

**Comment by AdrienPoupa** [comment]
Let's continue the discussion here: https://github.com/VinylMusicPlayer/VinylMusicPlayer/discussions/641

**Comment by soncaokim** [comment]
Hi @Octoton, do you plan to push this feature?

**Comment by soncaokim** [comment]
Closing this, the main branch deviated since and no update on the PR.

**Pull request closed without merging**